src/libtiled/compression.cpp

```diff
@@ -65,116 +66,159 @@ static void logZlibError(int error)
-                inflateEnd(&strm);
-                logZlibError(ret);
-                return QByteArray();
+        if (ret != Z_OK) {
+            logZlibError(ret);
+            return QByteArray();
         }
 
-        if (ret != Z_STREAM_END) {
-            int oldSize = out.size();
-            out.resize(oldSize * 2);
-
-            strm.next_out = (Bytef *)(out.data() + oldSize);
-            strm.avail_out = oldSize;
+        do {
+            ret = inflate(&strm, Z_SYNC_FLUSH);
+            Q_ASSERT(ret != Z_STREAM_ERROR);
+
+            switch (ret) {
+                case Z_NEED_DICT:
+                    ret = Z_DATA_ERROR;
+                    Q_FALLTHROUGH();
+                case Z_DATA_ERROR:
+                case Z_MEM_ERROR:
+                    inflateEnd(&strm);
+                    logZlibError(ret);
+                    return QByteArray();
+            }
+
+            if (ret != Z_STREAM_END) {
+                int oldSize = out.size();
+                out.resize(oldSize * 2);
+
+                strm.next_out = (Bytef *)(out.data() + oldSize);
+                strm.avail_out = oldSize;
+            }
         }
-    }
-    while (ret != Z_STREAM_END);
+        while (ret != Z_STREAM_END);
 
-    if (strm.avail_in != 0) {
-        logZlibError(Z_DATA_ERROR);
-        return QByteArray();
-    }
+        if (strm.avail_in != 0) {
+            logZlibError(Z_DATA_ERROR);
+            return QByteArray();
+        }
 
-    const int outLength = out.size() - strm.avail_out;
-    inflateEnd(&strm);
+        const int outLength = out.size() - strm.avail_out;
+        inflateEnd(&strm);
 
-    out.resize(outLength);
-    return out;
+        out.resize(outLength);
+        return out;
+    }
+    else
+    {
+        size_t const dSize = ZSTD_decompress(out.data(), out.size(), data.constData(), data.size());
+        if (ZSTD_isError(dSize)) {
+            qDebug() << "error decoding:" << ZSTD_getErrorName(dSize);
+            return QByteArray();
+        }
+        out.resize(dSize);
+        return out;
+    }
 }
 
-QByteArray Tiled::compress(const QByteArray &data, CompressionMethod method)
+QByteArray Tiled::compress(const QByteArray &data, const CompressionMethod method, const unsigned int compressionlevel)
 {
     if (data.isEmpty())
         return QByteArray();
 
-    QByteArray out;
-    out.resize(1024);
-    int err;
-    z_stream strm;
-    strm.zalloc = Z_NULL;
-    strm.zfree = Z_NULL;
-    strm.opaque = Z_NULL;
-    strm.next_in = (Bytef *) data.data();
-    strm.avail_in = data.length();
-    strm.next_out = (Bytef *) out.data();
-    strm.avail_out = out.size();
-
-    const int windowBits = (method == Gzip) ? 15 + 16 : 15;
-
-    err = deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits,
-                       8, Z_DEFAULT_STRATEGY);
-    if (err != Z_OK) {
-        logZlibError(err);
-        return QByteArray();
-    }
+    if(method==Zlib || method==Gzip)
+    {
```

Comment:
Coding style:
* Please put a space after `if`
* Please put spaces around `==`
* Please put the `{` on the same line as the `if` or `else`, also put `else` on the same line as the `}`